android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java

```diff
@@ -257,56 +259,80 @@ public void onDownloadStart(String url, String userAgent, String contentDisposit
       }
     });
 
-    return webView;
+    return wrapper;
   }
 
   @ReactProp(name = "javaScriptEnabled")
-  public void setJavaScriptEnabled(WebView view, boolean enabled) {
-    view.getSettings().setJavaScriptEnabled(enabled);
+  public void setJavaScriptEnabled(RNCWebView view, boolean enabled) {
+    view.getWebView().getSettings().setJavaScriptEnabled(enabled);
   }
 
   @ReactProp(name = "setBuiltInZoomControls")
-  public void setBuiltInZoomControls(WebView view, boolean enabled) {
-    view.getSettings().setBuiltInZoomControls(enabled);
+  public void setBuiltInZoomControls(RNCWebView view, boolean enabled) {
+    view.getWebView().getSettings().setBuiltInZoomControls(enabled);
   }
 
   @ReactProp(name = "setDisplayZoomControls")
-  public void setDisplayZoomControls(WebView view, boolean enabled) {
-    view.getSettings().setDisplayZoomControls(enabled);
+  public void setDisplayZoomControls(RNCWebView view, boolean enabled) {
+    view.getWebView().getSettings().setDisplayZoomControls(enabled);
   }
 
   @ReactProp(name = "setSupportMultipleWindows")
-  public void setSupportMultipleWindows(WebView view, boolean enabled){
-    view.getSettings().setSupportMultipleWindows(enabled);
+  public void setSupportMultipleWindows(RNCWebView view, boolean enabled){
+    view.getWebView().getSettings().setSupportMultipleWindows(enabled);
+  }
+
+  @ReactProp(name = "webViewKey")
+  public void setWebViewKey(RNCWebView view, String webViewKey) {
+    Map<String, RNCWebView> rncWebViewMap = RNCWebViewMapManager.INSTANCE.getRncWebViewMap();
+
+    if (rncWebViewMap.containsKey(webViewKey)) {
+      RNCWebView existingView = rncWebViewMap.get(webViewKey);
+
+      InternalWebView existingWebView = existingView.detachWebView();
```

Comment:
https://github.com/discord/react-native-webview/blob/11.18.1-discord-1/apple/RNCWebViewMapManager.m
https://github.com/discord/react-native-webview/blob/11.18.1-discord-1/apple/RNCWKWebViewMapManager.m

In the iOS implementation, we have a map for both the outer view group `RNCWebView` and also the inner WebView `WKWebView`. If I'm remembering correctly, depending on how the JS for the react components was implemented, it was possible for the old `RNCWebView` to get destroyed before the new one got created. In that case, we wouldn't be able to use the old `RNCWebView` to get the old `InternalWebView` via `detachWebView`. Instead, we would have to grab the old `InternalWebView` / `WKWebView` from a singleton map.

Is that possible on Android, too? In the test app, I think you can test this with the "toggle pages" button. That should let you test destroying the first `RNCWebVIew` without immediately getting a second `RNCWebView`. Then when you navigate away from the web view page to a different page and then back to the web view page, you should be able to still restore the old `InternalWebView`. For that, you might need a map for `InternalWebView`, but let me know what you find.

Reply:
> In the test app, I think you can test this with the "toggle pages" button. That should let you test destroying the first RNCWebVIew without immediately getting a second RNCWebView. Then when you navigate away from the web view page to a different page and then back to the web view page, you should be able to still restore the old InternalWebView. For that, you might need a map for InternalWebView, but let me know what you find.

I am able to restore the old InternalWebView with Toggle Pages.

> if I'm remembering correctly, depending on how the JS for the react components was implemented, it was possible for the old RNCWebView to get destroyed before the new one got created.

I am not sure how this is possible because this method is called 
https://github.com/discord/react-native-webview/pull/11/files/de8f049d35b3911134285a514add9418eceec7ff#diff-e9700631fc9fa5600d2e1ee70eb2d28aa2a1fbc695b85f99981972d3904ca485R807
when the view is being destroyed. We are also maintaining a reference to this view inside the hashmap so it shouldn't be destroyed

I'll double check on iOS

Reply:
> We are also maintaining a reference to this view inside the hashmap so it shouldn't be destroyed

I guess I was thinking that when the react component unmounts, we wouldn't hold onto the wrapper / RNCWebView, and instead, we'd only hold onto the InternalWebView / WKWebView. That way, we'd be holding on to only what's necessary without using too much memory (since it's fine if the `RNCWebView` gets destroyed when the react component unmounts as long as the `InternalWebView` stays alive).

In this code in the iOS `RNCWebView.m`,

```
- (void)removeFromSuperview
{
  bool keepWebViewInstance = _keepWebViewInstanceAfterUnmount && _webViewKey != nil;
  if (!keepWebViewInstance) {
    [self cleanUpWebView];
  }
  
  if (_webViewKey != nil) {
    NSMutableDictionary *sharedRNCWebViewDictionary= [[RNCWebViewMapManager sharedManager] sharedRNCWebViewDictionary];
    RNCWebView *rncWebView = sharedRNCWebViewDictionary[_webViewKey];
      
    // When this view is being unmounted, only remove the WKWebView from the superview
    // if this RNCWebView is the "active" view.
    if (rncWebView == self) {
      [self removeWKWebViewFromSuperView:self];
    }
  }

  [super removeFromSuperview];
}

```

, we remove the `RNCWebView` from the map when the react component unmounts. I think that part is different from the current Android implementation because it seems like we're keeping the Android `RNCWebView` in the map when the component unmounts.

> I'll double check on iOS

Sounds good! Let me know if you find anything different from what I described above

Reply:
> That way, we'd be holding on to only what's necessary without using too much memory (since it's fine if the RNCWebView gets destroyed when the react component unmounts as long as the InternalWebView stays alive).

I think the added memory here is marginal since it can be at most one wrapper view for each webViewKey. The wrapper view is also a single FrameLayout and I don't expect to see any reasonable performance gains.

I think the bigger question is consistency between iOS and Android - if there is no real need to maintain another map, I think it's easier to to maintain one map. iOS is slightly different in the sense that the RNCWebView is more complex so it might justify another map even though it may not be technically required.

Reply:
> I think the bigger question is consistency between iOS and Android - if there is no real need to maintain another map, I think it's easier to to maintain one map. iOS is slightly different in the sense that the RNCWebView is more complex so it might justify another map even though it may not be technically required.

Makes sense. The consistency is valuable to me to make it easier for us to apply similar changes to both platforms. I'm leaning toward keeping the iOS and Android implementations more consistent. I think the easiest way to do that today would be to make Android have two maps (to match the current iOS implementation): one for the outer webview and one for the inner webview. Alternatively, I'm also open to us making iOS only have one map to match what this Android PR is doing if that doesn't create other issues on iOS.

What do you think?

Reply:
I agree for consistency, given that iOS is somewhat battle-tested I'd leave that alone for now. I'll update Android and leave a comment that this is to remain consistent with iOS

Reply:
Sounds good!

Reply:
updated